### PR TITLE
ref(setup): Split setup into smaller functions into submodules

### PR DESCRIPTION
**Pull request by oberien**

### CODE REVIEW

File: windows-gaming-driver/src/config.rs [hunk 1]
```diff
@@ -47,19 +47,19 @@ pub struct MachineConfig {
...
 }
 
-#[derive(Deserialize, Serialize, Debug, Clone)]
+#[derive(Deserialize, Serialize, Debug, Clone, Default)]
 pub struct NetworkConfig {
     pub bridges: Vec<String>, // TODO: custom usernet
 }
 
-#[derive(Deserialize, Serialize, Debug, Clone)]
+#[derive(Deserialize, Serialize, Debug, Clone, Default)]
```
Comment (main--) [inline windows-gaming-driver/src/config.rs]: `StorageDevice`, `NetworkConfig` and `SambaConfig` shouldn't be `Default`. It doesn't make sense for any of them and it's not required either as they only appear in `Vec` and `Option`.

File: windows-gaming-driver/src/setup/vm.rs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    setup.floppy = Some(datadir.join("virtio-win.vfd").to_str().unwrap().to_owned());
+    true
+}
+
+pub fn memory(machine: &mut MachineConfig) -> bool {
+    // TODO: hugepages part 2
+    if machine.memory == "" {
+        // FIXME: validate this
+        machine.memory = ask::anything("How much memory would you like to assign to it?",
+                                              "be careful no validation LUL", |x| Some(x.to_owned()));
```
Comment (main--) [inline windows-gaming-driver/src/setup/vm.rs]: indent messed up

File: windows-gaming-driver/src/setup/vm.rs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+    // Stage 1: First boot and Windows setup
+
+    // TODO: if installed, call windows10-get-download-link, wget to a temporary location and go on
+    setup.cdrom = Some(ask::file("Please enter the path to your Windows ISO"));
+    setup.floppy = Some(datadir.join("virtio-win.vfd").to_str().unwrap().to_owned());
+    true
+}
+
+pub fn memory(machine: &mut MachineConfig) -> bool {
```
Comment (main--) [inline windows-gaming-driver/src/setup/vm.rs:63]: why are all these inner utility functions public? (affects other steps as well)

File: windows-gaming-driver/src/setup/wizard.rs [hunk 4]
```diff
@@ -1,307 +1,34 @@
...
+struct Wizard;
 
-        if let Some(cfg) = cfg {
-            machine = cfg.machine;
-            if let Some(cfg_setup) = cfg.setup {
-                setup = cfg_setup;
-            }
+impl Wizard {
+    fn run(&mut self, cfg: Option<Config>, cfg_path: &Path, workdir: &Path, datadir: &Path) {
+        let mut config = cfg.unwrap_or_default();
```
Comment (main--) [inline windows-gaming-driver/src/setup/wizard.rs:21]: i like this

File: windows-gaming-driver/src/setup/wizard.rs [hunk 5]
```diff
@@ -1,307 +1,34 @@
...
+        if config.setup.is_none() {
+            config.setup = Some(SetupConfig::default());
```
Comment (main--) [inline windows-gaming-driver/src/setup/wizard.rs:23]: the main reason why the old code kept the config parts separate is mostly this: We know there has to be a `SetupConfig` so we don't want to `unwrap` everywhere

I don't care much either way though

**Review by main-- — APPROVED** [review]
Not sure how I feel about dropping and re-acquiring stdin lock and udev context all the time.

I'm guessing you're trying to move towards some error_chain thing long-term as a `Result` is obviously more appropriate than a `bool` for all those steps (plus it avoids `expect` all over the place).

Also if you're gonna remove the whole point of the `Wizard` struct, don't just leave it there being useless.

File: windows-gaming-driver/src/setup/vm.rs (hunk 3 above)
Comment (oberien) [reply windows-gaming-driver/src/setup/vm.rs:63]: While not needed in this PR, this is in preparation for executing single steps.

File: windows-gaming-driver/src/setup/wizard.rs (hunk 5 above)
Comment (oberien) [reply windows-gaming-driver/src/setup/wizard.rs:23]: As I refactored the `run` method to be quite minimal now and basically only a grouping of functions (which may change in the future), I thought it'd be easier to read the saving of the config and see the config as such.
Additionally, it saves quite a few function calls (which would definitely not be inlined or so) and clones (you know, the CPU is quite slow when it comes to memcopy) :stuck_out_tongue:

**Comment by oberien** [comment]
The `Wizard` struct will become a wrapper around the individual functions, which can then be called by `main` (the function, not you) to execute single steps.
Did the rest.

**Pull request merged**